src/runtime_src/xrt/device/hal2.cpp

```diff
@@ -50,9 +51,16 @@ send_exception_message(const std::string& msg)
 namespace xrt_xocl { namespace hal2 {
 
 device::
-device(std::shared_ptr<operations> ops, unsigned int idx)
-  : m_ops(std::move(ops)), m_idx(idx), m_devinfo{}
+device(std::shared_ptr<operations> ops, unsigned int idx, const std::string& dll)
+  : m_ops{std::move(ops)}
+  , m_idx(idx)
+  , mFileName{std::move(dll)}
+  , m_devinfo{}
 {
+  // creating xrt device ahead so that other functions can
+  // be called using the same
+  open();
```

Comment:
This fundamentally changes when a device is opened.
I am hoping this is okay, but there must be some reason why it was deferred earlier?

I am hesitant with this change.  Can we wrap getting the xrt_core::device* in a function that returns xrt_core::device*.  Something like:

```
xrt_core::device*
get_core_device_raw() 
{
  if (auto cdev = m_handle.get_handle().get())  // xrt_core::device*
    return cdev;

  // if xrt::device has not be created it is wrong to call this function
  throw std::runtime_error("Internal error: device has not been opened");
}
```
All instances of m_core_dev_hdl->zzz() to be replaced with get_core_device_raw()->zzz()